// @midwayjs/redis/dist/configuration.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisConfiguration = void 0;
const core_1 = require("@midwayjs/core");
const manager_1 = require("./manager");
let RedisConfiguration = class RedisConfiguration {
    async onReady(container) {
        await container.getAsync(manager_1.RedisServiceFactory);
    }
    async onStop(container) {
        const factory = await container.getAsync(manager_1.RedisServiceFactory);
        await factory.stop();
    }
    async onHealthCheck(container) {
        const factory = await container.getAsync(manager_1.RedisServiceFactory);
        const clients = factory.getClients();
        // find status not ready
        let clientName;
        for (const [name, instance] of clients) {
            if (instance.status !== 'ready' && !factory.isLowPriority(name)) {
                clientName = name;
                break;
            }
        }
        return {
            status: !clientName,
            reason: clientName ? `redis client "${clientName}" is not ready` : '',
        };
    }
};
RedisConfiguration = __decorate([
    (0, core_1.Configuration)({
        namespace: 'redis',
        importConfigs: [
            {
                default: {
                    redis: {},
                },
            },
        ],
    })
], RedisConfiguration);
exports.RedisConfiguration = RedisConfiguration;
