// minidev/assets/builder-debug-utils/lib/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuilderDebugContext = void 0;
const path_1 = require("path");
class BuilderDebugContext {
    constructor(config) {
        this.assetsQueryUrl = '';
        this.injectCodeTempPath = '';
        this.offlineDirectoryPath = '';
        this.readonlyDirectoryPath = '';
        this.assetsMapFilename = '';
        this.assetsQueryUrl = config.assetsQueryUrl;
        this.injectCodeTempPath = config.injectCodeTempPath;
        this.offlineDirectoryPath = config.offlineDirectoryPath;
        this.readonlyDirectoryPath = config.readonlyDirectoryPath;
        this.assetsMapFilename = config.assetsMapFilename;
        this.boatmanFilename = config.boatmanFilename;
    }
    getAssetsMapFilePath() {
        return (0, path_1.join)(this.offlineDirectoryPath, this.assetsMapFilename);
    }
    getBoatmanFilePath(target) {
        return (0, path_1.join)(this.offlineDirectoryPath, this.boatmanFilename[target]);
    }
    getReadonlyAssetsMapFilePath() {
        return (0, path_1.join)(this.readonlyDirectoryPath, this.assetsMapFilename);
    }
    getReadonlyBoatmanFilePath(target) {
        return (0, path_1.join)(this.readonlyDirectoryPath, this.boatmanFilename[target]);
    }
}
exports.BuilderDebugContext = BuilderDebugContext;
